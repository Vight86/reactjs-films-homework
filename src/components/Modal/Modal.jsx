import React from 'react';
import propTypes from 'prop-types';
import { createPortal } from 'react-dom';
import Button from '../Button/index';
import Preloader from '../Preloader/index';
import styles from './Modal.scss';

const Modal = ({ isModalOpened, movieTrailerKey, handleModalToggle }) => {
  if (!isModalOpened) return null;

  const modal = (
    <div className={styles.modalWrapper}>
      <div className={styles.modalVideo}>
        <Button
          className="closeButton_inModal"
          onClick={() => handleModalToggle(null, isModalOpened)}
        >
          &times;
        </Button>
        {
          movieTrailerKey
          && (
            <iframe
              className={styles.modalIframe}
              src={`https://www.youtube.com/embed/${movieTrailerKey}`}
              title="YouTube video player"
              frameBorder="0"
              allow="accelerometer; autoplay; clipboard-write; encrypted-media; gyroscope; picture-in-picture"
              allowFullScreen
            />
          )
        }
        <Preloader className="inModal" />
      </div>
    </div>
  );

  return createPortal(modal, document.getElementById('modal'));
};

Modal.propTypes = {
  isModalOpened: propTypes.bool.isRequired,
  movieTrailerKey: propTypes.string,
  handleModalToggle: propTypes.func.isRequired,
};

Modal.defaultProps = {
  movieTrailerKey: null,
};

export default Modal;
