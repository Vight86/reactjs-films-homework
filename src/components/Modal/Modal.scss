@import '../../utils/variables';

.modalWrapper {
  position: fixed;
  z-index: 1000;
  top: 0;
  left: 0;
  display: flex;
  width: 100%;
  height: 100%;
  background-color: transparentize($color: $color-dark, $amount: 0.1);
  overflow-y: auto;
}

.modalVideo {
  position: relative;
  display: flex;
  width: 70%;
  align-items: center;
  justify-content: center;
  border: 5em solid transparent;
  margin: auto;

  &::before {
    position: relative;
    display: block;
    padding-top: 56.3%;
    content: '';
  }
}

.modalIframe {
  position: absolute;
  z-index: 20;
  top: 0;
  left: 0;
  width: 100%;
  height: 100%;
}
