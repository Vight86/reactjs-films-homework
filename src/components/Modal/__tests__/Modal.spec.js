import React from 'react';
import ShallowRenderer from 'react-test-renderer/shallow';
import Modal from '../index';

let mockHandleModalToggle;
let renderer;
let result;

beforeEach(() => {
  const container = document.createElement('div');
  container.id = 'modal';
  document.body.appendChild(container);

  mockHandleModalToggle = jest.fn(() => 'test');

  renderer = new ShallowRenderer();
  renderer.render(
    <Modal
      isModalOpened={false}
      movieTrailerKey={null}
      handleModalToggle={mockHandleModalToggle}
    />,
  );
  result = renderer.getRenderOutput();
});

afterEach(() => {
  mockHandleModalToggle.mockClear();
  renderer = null;
  result = null;
});

describe('render Modal component', () => {
  it('render initial state correctly', () => {
    expect(result).toMatchSnapshot();
  });

  it('render modal open state correctly', () => {
    renderer.render(
      <Modal
        isModalOpened
        movieTrailerKey="test"
        handleModalToggle={mockHandleModalToggle}
      />,
    );
    result = renderer.getRenderOutput();
    expect(result).toMatchSnapshot();
  });

  it('render handleModalToggle correctly', () => {
    renderer.render(
      <Modal
        isModalOpened
        movieTrailerKey="test"
        handleModalToggle={mockHandleModalToggle}
      />,
    );
    result = renderer.getRenderOutput();
    const closeButton = result.children.props.children.props.children[0];
    closeButton.props.onClick();
    expect(mockHandleModalToggle).toHaveBeenCalled();
  });
});
