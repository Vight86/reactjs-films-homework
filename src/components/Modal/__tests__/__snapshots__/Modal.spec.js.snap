// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render Modal component render initial state correctly 1`] = `null`;

exports[`render Modal component render modal open state correctly 1`] = `
Object {
  "$$typeof": Symbol(react.portal),
  "children": <div
    className="modalWrapper"
  >
    <div
      className="modalVideo"
    >
      <Button
        className="closeButton_inModal"
        onClick={[Function]}
      >
        Ã—
      </Button>
      <iframe
        allow="accelerometer; autoplay; clipboard-write; encrypted-media; gyroscope; picture-in-picture"
        allowFullScreen={true}
        className="modalIframe"
        frameBorder="0"
        src="https://www.youtube.com/embed/test"
        title="YouTube video player"
      />
      <Preloader
        className="inModal"
      />
    </div>
  </div>,
  "containerInfo": <div
    id="modal"
  />,
  "implementation": null,
  "key": null,
}
`;
